x/cron/keeper/keeper.go

```diff
@@ -0,0 +1,264 @@
+}
+
+func (k *Keeper) GetScheduleCount(ctx sdk.Context) int32 {
+	return k.getScheduleCount(ctx)
+}
+
+func (k *Keeper) StoreSchedule(ctx sdk.Context, schedule types.Schedule) {
+	if !k.scheduleExists(ctx, schedule.Name) {
+		k.changeTotalCount(ctx, 1)
+	}
+	k.storeSchedule(ctx, schedule)
+}
+
+func (k *Keeper) getSchedulesReadyForExecution(ctx sdk.Context) []types.Schedule {
+	params := k.GetParams(ctx)
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+	count := uint64(0)
+
+	res := make([]types.Schedule, 0)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var schedule types.Schedule
+		k.cdc.MustUnmarshal(iterator.Value(), &schedule)
+
+		if k.intervalPassed(ctx, schedule) {
+			res = append(res, schedule)
+			count++
+
+			if count >= params.Limit {
+				k.Logger(ctx).Info("limit of schedule executions per block reached")
+				return res
+			}
+		}
+	}
+
+	return res
+}
+
+// executeSchedule executes all msgs in a given schedule and changes LastExecuteHeight
+// if at least one msg execution fails, rollback all messages
+func (k *Keeper) executeSchedule(ctx sdk.Context, schedule types.Schedule) error {
+	// Even if contract execution returned an error, we still increase the height
+	// and execute it after this interval
+	schedule.LastExecuteHeight = uint64(ctx.BlockHeight())
+	k.storeSchedule(ctx, schedule)
+
+	subCtx, commit := ctx.CacheContext()
+
+	for idx, msg := range schedule.Msgs {
+		executeMsg := wasmtypes.MsgExecuteContract{
+			Sender:   k.accountKeeper.GetModuleAddress(types.ModuleName).String(),
+			Contract: msg.Contract,
+			Msg:      []byte(msg.Msg),
+			Funds:    sdk.NewCoins(),
+		}
+		_, err := k.WasmMsgServer.ExecuteContract(sdk.WrapSDKContext(subCtx), &executeMsg)
+		if err != nil {
+			ctx.Logger().Info("executeSchedule: failed to execute contract msg",
+				"schedule_name", schedule.Name,
+				"msg_idx", idx,
+				"msg_contract", msg.Contract,
+				"msg", msg.Msg,
+				"error", err,
+			)
+			return err
+		}
+	}
+
+	// only save state if all the messages in a schedule were executed successfully
+	commit()
+	return nil
+}
+
+func (k *Keeper) storeSchedule(ctx sdk.Context, schedule types.Schedule) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+
+	bzSchedule := k.cdc.MustMarshal(&schedule)
+	store.Set(types.GetScheduleKey(schedule.Name), bzSchedule)
+}
+
+func (k *Keeper) removeSchedule(ctx sdk.Context, name string) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+
+	store.Delete(types.GetScheduleKey(name))
+}
+
+func (k *Keeper) scheduleExists(ctx sdk.Context, name string) bool {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+	return store.Has(types.GetScheduleKey(name))
+}
+
+func (k *Keeper) intervalPassed(ctx sdk.Context, schedule types.Schedule) bool {
+	return uint64(ctx.BlockHeight()) > (schedule.LastExecuteHeight + schedule.Period)
```

Comment:
wdyt about doing something fancy like `(ctx.BlockHeight() % schedule.Period) == 0`

Reply:
looks fancy, but in case a task queue size more than the "limit". Lets say the limit is 3, but we have 4 tasks with period 3, then we constantly drop a 4th task

Reply:
This can happen, but I meant this limit for situation where too many schedules are ready on one block.
So situation where there are too many schedules overall can happen, but it should not be feasible, since it's government decisions that add schedules

Reply:
Looked better into it, and agree with @swelf19, current solution better for me since it's moves the unhandled load on the next block, your solution moves the load onto next period, so if the period is overloaded, it will happen forever

Reply:
@swelf19 what?.. what does queue size have to do with checking whether enough blocks have passed?